### PR TITLE
Complete the menu

### PR DESCRIPTION
**Pull request by xsy1996**
Add "Use Cached ISO" opton to main menu. which will displays the cache directory nested, until reach the iso file. change the structure of cache directory will change the hierarch of menu at the same time.

Signed-off-by: April Xu <shiyux@google.com>

### CODE REVIEW

**Comment by xsy1996** [comment]
This is a screen recording of test
cinnamon-20200731-3

**Comment by rminnich** [comment]
This is great. It is something I don't think I'd ever figure out in a million years. Wow.

**Comment by rminnich** [comment]
I did run it and enter ab.dc for the URL and the program exited, it would be nice if it was a bit more gracious.

Also do I have a way to see what bookmarks there are.

**Comment by xsy1996** [comment]
Sorry I think the reason enter ab.dc for the URL and the program exited is because I mistakely only print the error message  in the verbose mode. Actually the previous version does have a way to see what bookmarks there are. But the menu may looks complicated. So Prachi suggesst to just let user input the name of iso they want. And if we can find a matched name in the bookmark, we use the book mark, elsewise we ask for a download link. The biggest problem of menu is  termui's ui is not scrollable so the length of menu is limited. so if we want to display the bookmark, we may have to split the bookmark into small pieces and display them level by level

**Comment by xsy1996** [comment]
demo of entering invalid download link
wrong_download_link

File: cmds/webboot/webboot.go [hunk 1]
```diff
@@ -51,69 +47,60 @@ func (d *DownloadOption) exec(uiEvents <-chan ui.Event) (menu.Entry, error) {
...
 	fpath := filepath.Join("/tmp", filename)
-	if err = download(link, fpath); err != nil {
-		return nil, err
-	}
-
-	var iso *ISO = &ISO{
-		label: filename,
-		path:  fpath,
+	// if download link is not valid, ask again until the link is rights
+	for err = download(link, fpath); err != nil; err = download(link, fpath) {
```
Comment (pallaud) [inline cmds/webboot/webboot.go:51]: Why do we need to do `err = download(link, fpath)` twice? Hm would it work to do the loop this way

```
derr := download(link, fpath)
for derr != nil {
  // display derr
 // get input
  derr = download(link, fpath)
}
```

File: cmds/webboot/webboot.go [hunk 2]
```diff
@@ -108,12 +132,16 @@ func download(URL, fPath string) error {
...
-	log.Println(*v)
 	if *v {
 		verbose = log.Printf
 	}
 
 	entries := []menu.Entry{
+		&DirOption{
+			label: "Use Cached ISO",
+			// todo: replace ./testdata with cache directory
+			path: "./testdata",
```
[the remark this answers is not in the document]
Comment (pallaud) [reply cmds/webboot/webboot.go:102]: Can we do this in this PR itself? It shouldn't be too hard to do so its preferable to do it now and not hardcode testdata.

File: pkg/menu/menu.go [hunk 3]
```diff
@@ -84,15 +84,20 @@ func processInput(introwords string, location int, wid int, ht int, isValid vali
 				warning.Text = ""
 				ui.Render(warning)
 			}
-			input.Text += k
+			if k == "<Space>" {
+				input.Text += " "
+			} else if k[0:1] != "<" {
```
Comment (pallaud) [inline pkg/menu/menu.go:89]: I'm a little confused, why do we need this check?

Comment (xsy1996) [reply pkg/menu/menu.go:89]: because for example we hit left arrow, the termui's ui.PollEvents() will push a string"<Left>" to the channal. and all other special key like arrow, F1,etc on the keyboard will be treat in this way. So if the key is not begin at '<' it means it's a alphabet or number or .

File: cmds/webboot/webboot.go (hunk 1 above)
Comment (xsy1996) [reply cmds/webboot/webboot.go:51]: Ok that would works too thanks. the reason I write download(link, fpath) twice here because the for loop's first statement is a initialization, second is the judgement; third is the statement will run at the begin of every loop.

File: cmds/webboot/webboot.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (xsy1996) [reply cmds/webboot/webboot.go:102]: sorry the revise of PR haven't done yet, so I do not call you to review. I plan to fix this part in the next commit

File: cmds/webboot/webboot_test.go [hunk 4]
```diff
@@ -104,15 +105,46 @@ func TestDownloadOption(t *testing.T) {
...
+				t.Errorf("Fail to execute option (%q)'s exec(): %+v", entry.Label(), err)
+				break
+			}
+		} else if iso, ok := entry.(*ISO); ok {
+			if iso.label != wanted.label || iso.path != wanted.path {
+				t.Errorf("Get wrong chosen iso. get %+v, want %+v", iso, wanted)
+			}
+			break
+		} else {
+			t.Errorf("Unknow type, get a %T of entry %+v", entry, entry)
```
Comment (pallaud) [inline cmds/webboot/webboot_test.go:146]: Nit: `t.Errorf("Unknown type. got entry of type %T, wanted DirOption or ISO", entry)`

File: cmds/webboot/utils.go [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"net/url"
+	"os"
+	"strings"
+)
+
+// WriteCounter counts the number of bytes written to it. It implements to the io.Writer
```
Comment (pallaud) [inline cmds/webboot/utils.go:13]: Could you move the progress bar change to another PR? Lets try and keep each PR focused on 1 thing.

File: cmds/webboot/Readme.md [hunk 6]
```diff
@@ -0,0 +1,8 @@
+### Usage
+The webboot program would do the following:
+ - Present a menu with the existing cached distro options
+ - If the user wants a distro that is not cached, they can download an ISO 
+ - After the user decides on an ISO, boot it.
+
+### Test
+Our UI uses a package called Termui. Termui will parse the standard input into keyboard events and insert them into a channel, then from which the Termui get it's input.  For implement a unattended test, I manually build a series of keyboard events that reperesent my intented input for test, and insert them into a channel. Then I replace the original input channel with my channel in the test. So the go test could run a test of ui automatically.
```
Comment (pallaud) [inline cmds/webboot/Readme.md:8]: Could you give an example here maybe? Say something like "See TestDownloadOption for an example. We use `go pressKey(uiEvents, input)` to set up the channel..."

File: cmds/webboot/webboot.go [hunk 7]
```diff
@@ -51,69 +48,65 @@ func (d *DownloadOption) exec(uiEvents <-chan ui.Event) (menu.Entry, error) {
...
 func main() {
 	flag.Parse()
-	log.Println(*v)
 	if *v {
 		verbose = log.Printf
 	}
 
+	cachedDir := *dir
+	if cachedDir == "" {
+		// todo: find the cache directory by check block devices
```
[the remark this answers is not in the document]
Comment (xsy1996) [reply cmds/webboot/webboot.go:101]: Oh yeah, that would be better thanks.

File: pkg/menu/menu.go [hunk 8]
```diff
@@ -84,15 +84,20 @@ func processInput(introwords string, location int, wid int, ht int, isValid vali
 				warning.Text = ""
 				ui.Render(warning)
 			}
-			input.Text += k
+			if k == "<Space>" {
```
[the remark this answers is not in the document]
Comment (xsy1996) [reply pkg/menu/menu.go:87]: I think <Space> is just a special input character. if i make it another switch statement, I should repeat the logic like checking warning text etc.

File: cmds/webboot/utils.go (hunk 5 above)
Comment (xsy1996) [reply cmds/webboot/utils.go:13]: Ok sure

File: pkg/menu/menu.go (hunk 3 above)
Comment (xsy1996) [reply pkg/menu/menu.go:89]: ok, will do
